walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,100 @@
+    Include optional attendees if at least one time slot can be found
+    Otherwise return the time slots that fit just the mandatory attendees.
+*/
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Initialize a time slot (the whole day)
+    List<TimeRange> timeSlots = new ArrayList<>();
+    timeSlots.add(TimeRange.fromStartDuration(0, 24 * 60));
+
+    // Find time slots for required attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    // Save the result in timeSlotsForRequired
+    // in case no time slot can be found when optional attendees are included
+    List<TimeRange> timeSlotsForRequired = new ArrayList<TimeRange>(timeSlots);
+
+    // Check for the availability of optional attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getOptionalAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    if (timeSlots.isEmpty() && !request.getAttendees().isEmpty()) {
+      return timeSlotsForRequired;
+    }
+    return timeSlots;
+  }
+
+  // Iterate through all event to check
+  // if a request attendee is attending a certain event
+  public List<TimeRange> queryForEvent(Collection<Event> events,
+      List<TimeRange> timeSlots, Collection<String> attendees) {
+    for (Iterator<Event> i = events.iterator(); i.hasNext();) {
+      Event event = i.next();
+
+      // Iterate through attendees to check the events each of them are attending
+      for (Iterator<String> j = attendees.iterator(); j.hasNext();) {
+        String attendee = j.next();
+        if (event.getAttendees().contains(attendee)) {
+
+          // Iterate through all time periods to see
+          // if any period overlaps with an event the attendees have
+          for (ListIterator<TimeRange> k = timeSlots.listIterator(); k.hasNext();) {
+            TimeRange currTime = k.next();
+            TimeRange eventTime = event.getWhen();
+
+            // Removes the time period when completely covered by event time
+            if (currTime.equals(eventTime) || eventTime.contains(currTime)) {
+              k.remove();
+
+            // Case B: |-----B------|
+            // Event:      |---|
+            //         |-1-|   |-2--|
+            // Split time period from B into 1 and 2
+            } else if (currTime.contains(eventTime)) {
+              k.remove();
+              k.add(TimeRange.fromStartEnd(currTime.start(), eventTime.start(), false));
+              k.add(TimeRange.fromStartEnd(eventTime.end(), currTime.end(), false));
+            
+            // Case C: |--C---|
+            // Event:      |------|
+            //         |-1-|
+            // Shorten time period from C to 1
```

Comment:
1) Good use of ASCII art to clarify the algorithm's different scenarios, I like it!
2) Do 'slot' and 'period' mean the same thing? If yes, I'd suggest standardizing for clarity.
3) I like how one timeSlot is named "Event". To be complete, what would be the name of the other two timeslots (lines 83 and 85)?

Reply:
2) Fixed. 3) I changed Case C and event to their corresponding variable name so it matches with the code better.